Separate workflow/context with different loggers

kbrock opened the pull request:
## Goal

The logging is global to all of floe. This means we do not associate certain logging information with each user/run.

Setting up a per execution log allows us to isolate the log per running instance and display it to the end user.

## Implementation

The `Context` is a per execution concept. It contains the user's information (in execution), the user's credentials, and the runtime context. (including credentials in context has conflated the db context and the runtime context)

Since we want a per execution logger, I added logger to the `Context`.

## Alternative

The `Workflow` currently contains the states and the runtime `Context`. The suggestion was to add the `logger` to the `Workflow`. Unfortunately, this does not work because while the various components have access to the `Context`, they do not have access to the `Workflow`.

So I propose we stay the course and use `Context#logger`

## Extracted

pulled out the following PRs:

- [ ] https://github.com/ManageIQ/floe/pull/289

kbrock commented:
update:
- rebased to handle changes to cli

think this is still part of discussions

Fryguy commented:
Overall I like the concept, but I think of context as more of the "serializable" part of the workflow, so it feels like the wrong place to put the logger.  I think workflow itself feels like the more appropriate place. @agrare @kbrock thoughts?

Fryguy commented on lib/floe/cli.rb:
```diff
@@ -25,11 +26,25 @@ def run(args = ARGV)
+          {}
+        else
+          workflows.each_with_object({}) do |workflow, h|
+            workflow.context.logger = Logger.new(output = StringIO.new)
```
I'm not sure I like this.  I like seeing the live logs when using exe/floe, because I can see what state its in as it's running.

Fryguy replied:
Oh wait, I misread the code.  I see that passing segment output will trigger this section otherwise it works normally. 👍

kbrock commented:
@Fryguy Yes, @agrare feels the same way as you do.

- `source.asl` -- source code.
- AST -- source code loaded into memory. This can be the same across multiple executions.
- Execution Environment
  - state input `Context#state.input` / `Context#input`
  - credentials `Context#credentials` (recent change)
  - execution history `Context#state_history`
  - logging (was suggesting `Context.logger` but up to you)

I was pushing that `Workflow` would only be the `AST`, but looks like we have linked `context` in there. I did try and pull `context` out, but it didn't work so well.

Maybe we want a different object to group `context` (really only need the current state), `credentials`, `logger`

kbrock commented:
update:
- rebase (conflict)

kbrock commented:
update:
- fix code climate (make Cli#run smaller)

update:
- fix refactor amuck.

kbrock commented:
update:
- rebase master

agrare commented:
Other than the logger belonging to Context I think this looks good, I would have thought logger lived with a Workflow

kbrock commented:
WIP: going to find another home for the logger. Possibly pass in a RuntimeContext, similar to an ImageProcessor context with Context State, credentials, and logger

kbrock commented:
un-WIP

Logistically, `Context#logger` works much better than `Workflow#logger`. The latter caused too many extra variables to get passed around, specifically in the `Runner`.

Philosophically, I feel the `Context` holds all execution and user specific state, while the `Workflow` holds no state and is valid across executions. By this definition, the logger belongs with the `Context`.

Elephant: Wihle I don't like the fact that `Workflow` currently has a reference to `Context`, I'm waiting for `wait` to settle down before suggesting any further change there.

Fryguy commented on lib/floe/cli.rb:
```diff
@@ -82,5 +90,22 @@ def parse_options!(args)
+        {}
+      else
+        workflows.each_with_object({}) do |workflow, h|
+          workflow.context.logger = Logger.new(output = StringIO.new)
```
Can you explain this string io part?  I don't understand this bit.  Is this meant to be like a NullLogger pattern (if so we should use an actual NullLogger like we do in other gems)?

Fryguy commented on lib/floe/cli.rb:
```diff
@@ -18,19 +19,25 @@ def run(args = ARGV)
-      Floe::Workflow.wait(workflows, &:run_nonblock)
+      output_streams = create_loggers(workflows, opts[:segment_output])
+
+      puts "Checking #{workflows.count} workflows..."
```
This won't be in the "real" application logs now though, right (since this is moved to cli, which is a dev tool)?  That was my concern.  Maybe it's not necessary in the application logs?

kbrock replied:
this is the cli. so yes, it will only go to stdout.
All entries in this file go to stdout via `puts`

Fryguy replied:
I guess what I'm asking is before this change, this particular log line went to the application logs and now it doesn't.  I'm not sure how useful that is - maybe @agrare has opinions here.

agrare replied:
Could we set the Floe.logger to stdout (if it isn't already) and
1. Keep the .wait logger the way it was
2. Use that instead of puts in cli.rb

kbrock replied:
Most logging we have is feedback for individual workflows (mostly step start/complete)

The cli uses `State.wait` to orchestrate which workflows are run. I think of this as part of the command line interface. So going away from stdout seems strange to me. (asid: I do want to refactor this method to go even based, but that is a future action)

So is the thought is to convert all cli `puts` statements to log statements?

kbrock commented:
update:
- removed change to specs (no longer convert state.run to workflow.run)

kbrock commented on lib/floe/workflow.rb:
```diff
@@ -110,9 +108,12 @@ def run_nonblock
       result = current_state.run_nonblock!(context)
       return result if result != 0
 
+      end_state
```
Yea, `result != 0` means the state has not finished yet. Something like a `Wait` or `Task` that has not completed yet.

In the future, it will call into `run_nonblock`

But I think I'll take this out and revisit what I'm trying to do here. Thanks

Fryguy replied:
oh i see it's the count of incomplete ones?  Interesting - Maybe result is just a bad variable name and something like "remaining_something" would be clearer.

kbrock commented:
update:
- output changed from `puts` to `logger`
- removed a number of extra changes
- drop start_state/end_state

update:
- added logging ()'s (code climate)

update:
- extra credentials to own method (code climate)

update:
- add logging ()'s (code climate - 1 remaining)

kbrock commented:
update:
- extracted cli refactoring to own PR
- rebased upon changes

Fryguy reviewed (approved):
LGTM - also want @agrare to review

kbrock commented:
@agrare You ok with this, or you want me to pull off one more PR and iterate a little more slowly.

I guess we could better partition runtime state and pull context out of there, but was hoping for that to be a future PR

agrare reviewed (approved):
I'm okay with this with the understanding that in the future we're going to split out our runtime context (lowercase c) from the Context :tm: object that is presented to the state machine
